Fixed #28142 -- Prevented is_safe_url from crashing when provided an invalid IPv6 URL.

UmanShahzad opened the pull request:
As per https://code.djangoproject.com/ticket/28142

Please let me know if this doesn't actually match the requirements or if a separate test is needed for this.

charettes commented on django/utils/http.py:
```diff
@@ -302,7 +302,10 @@ def _urlparse(url, scheme='', allow_fragments=True):
+    try:
+        splitresult = _urlsplit(url, scheme, allow_fragments)
+    except ValueError:
+        return False
```
This changes the return value of `_urlparse`. I think it would be more appropriate to catch the `ValueError` in `_is_safe_url` instead and return `False` there.

UmanShahzad replied:
Yes, that's more appropriate, will change!

charettes commented:
@UmanShahzad for the test, adding a line in [`bad_urls` should do](https://github.com/django/django/blob/1ebd295082bb0e6b230cf3bc39fd04bee71c2bd7/tests/utils_tests/test_http.py#L101).

UmanShahzad commented:
@charettes I've moved it as you said and added some invalid IPv6 URLs for testing.

One issue, and I'm not that knowledgeable with IPv6 so I need confirmation:

```
            '[2001:db8:a0b:12f0::1:21',
            '2001:db8:a0b:12f0::1]:21',
            '[2001:db8:a0b:12f0::1%25eth0:21',
            '2001:db8:a0b:12f0::1%25eth0]:21',
            '[2001:db8:a0b:12f0::1%eth0:80',
            '2001:db8:a0b:12f0::1%eth0]:80'
```

The above fail the tests (i.e. appear as valid IPv6 URLs) but I removed them because I'm not sure if they are technically valid or not. Can you give me some insight on that? If they should be included then some more work needs to be done, as far as I can tell.

timgraham commented:
I think only one invalid case needs to be tested as we're only testing the catching of `ValueError`, not that `_urlparse()` (mostly copied from Python's standard library) behaves properly.

UmanShahzad commented:
@timgraham That sounds fine to me! I left a single case (well really 2, but they check for both parts of the relevant conditional below:

```python
        if (('[' in netloc and ']' not in netloc) or
                (']' in netloc and '[' not in netloc)):
            raise ValueError("Invalid IPv6 URL")
```
).

Hope that's fine. :)
